src/services/request-authorizer.ts

```diff
@@ -0,0 +1,135 @@
+};
+
+const GLOB = '[^/]+';
+const AUTH_WHITELIST = [
+  '/login',
+  '/access-denied',
+  '/resident/*',
+  '/resident/*/upload',
+  '/resident/*/confirmation',
+].map((str) => new RegExp(`^${str.replace('*', GLOB)}$`));
+
+export interface RequestAuthorizerCommand {
+  path: string;
+  cookieHeader?: string;
+  serverSide: boolean;
+}
+
+export type RequestAuthorizerResponse =
+  | {
+      success: true;
+      user?: User;
+    }
+  | {
+      success: false;
+      redirect?: string;
+    };
+
+export interface RequestAuthorizerDependencies {
+  secret: string;
+  cookieName: string;
+  environmentKey: EnvironmentKey;
+  authGroups: Record<string, string>;
+  authWhitelist: RegExp[];
+}
+
+const defaultDependencies = {
+  secret: process.env.HACKNEY_JWT_SECRET as string,
+  cookieName: process.env.RUNTIME_HACKNEY_COOKIE_NAME as string,
+  environmentKey: process.env.RUNTIME_HOST_ENV as EnvironmentKey,
+  authGroups: authGroupsJson[process.env.RUNTIME_HOST_ENV as EnvironmentKey],
+  authWhitelist: AUTH_WHITELIST,
+};
+
+export class RequestAuthorizer {
+  private secret: string;
+  private cookieName: string;
+  private environmentKey: string;
+  private authGroups: Record<string, string>;
+  private authWhitelist: RegExp[];
+
+  constructor(deps: Partial<RequestAuthorizerDependencies> = {}) {
+    this.secret = deps.secret ?? defaultDependencies.secret;
+    this.cookieName = deps.cookieName ?? defaultDependencies.cookieName;
+    this.environmentKey =
+      deps.environmentKey ?? defaultDependencies.environmentKey;
+    this.authGroups = deps.authGroups ?? defaultDependencies.authGroups;
+    this.authWhitelist =
+      deps.authWhitelist ?? defaultDependencies.authWhitelist;
+  }
+
+  public execute(command: RequestAuthorizerCommand): RequestAuthorizerResponse {
+    const secureEnvironment =
+      command.serverSide && this.environmentKey !== 'dev';
+
+    const user = this.authoriseUser(secureEnvironment, command.cookieHeader);
+
+    if (this.pathIsWhitelisted(command.path)) return { success: true, user };
+
+    if (!user) {
+      const redirect = encodeURIComponent(command.path || '/');
+      const url = `/login?redirect=${redirect}`;
+
+      return { success: false, redirect: url };
+    }
+
+    if (!this.userIsInValidGroup(user)) {
+      return { success: false };
+    }
+
+    return { success: true, user };
+  }
+
+  private pathIsWhitelisted(path: string): boolean {
+    const { pathname } = new URL(path, 'http://hackney.gov.uk');
+    return this.authWhitelist.some((regex) => regex.test(pathname));
+  }
+
+  private userIsInValidGroup(user: User): boolean {
+    return Object.values(this.authGroups).some((group) =>
+      user.groups.includes(group)
+    );
+  }
+
+  // private deleteSession(res: ServerResponse): void => {
+  //   res.setHeader(
+  //     'Set-Cookie',
+  //     cookie.serialize(cookieName, '', {
+  //       maxAge: -1,
+  //       domain: '.hackney.gov.uk',
+  //     })
+  //   );
+  // };
```

Comment:
Should this be removed?

Reply:
I think it's a useful snippet for when we need to do logout